File: Platform/infra/http/interfaces/dotnet/BaseHttpClient.cs
using System.Text.Json;

namespace Coyote.Infra.Http;

/// <summary>
/// HTTP request implementation
/// </summary>
public class HttpRequest : IHttpRequest
{
    public string Url { get; set; } = string.Empty;
    public HttpMethod Method { get; set; } = HttpMethod.Get;
    public string? Body { get; set; }
    public IDictionary<string, string> Headers { get; } = new Dictionary<string, string>();
    public int? TimeoutMs { get; set; }
    public string? ClientCertPath { get; set; }
    public string? ClientKeyPath { get; set; }
    public string? CACertPath { get; set; }
    public bool VerifyPeer { get; set; } = true;
    public bool FollowRedirects { get; set; } = true;

    public void SetJsonBody<T>(T content)
    {
        Body = JsonSerializer.Serialize(content);
        SetHeader("Content-Type", "application/json");
    }

    public void SetHeader(string name, string value)
    {
        Headers[name] = value;
    }
}

/// <summary>
/// HTTP response implementation
/// </summary>
public class HttpResponse : IHttpResponse
{
    public int StatusCode { get; init; }
    public string Body { get; init; } = string.Empty;
    public IReadOnlyDictionary<string, string> Headers { get; init; } = new Dictionary<string, string>();
    public bool IsSuccess => StatusCode >= 200 && StatusCode < 300;
    public string? ErrorMessage { get; init; }

    public string? GetHeader(string name)
    {
        return Headers.FirstOrDefault(h => 
            string.Equals(h.Key, name, StringComparison.OrdinalIgnoreCase)).Value;
    }

    public T? GetContent<T>()
    {
        if (string.IsNullOrEmpty(Body))
            return default;

        try
        {
            return JsonSerializer.Deserialize<T>(Body);
        }
        catch (JsonException)
        {
            return default;
        }
    }
}

/// <summary>
/// Base HTTP client implementation with common functionality
/// </summary>
public abstract class BaseHttpClient : ICoyoteHttpClient
{
    protected readonly HttpClientOptions _options;
    protected bool _disposed;

    protected BaseHttpClient(HttpClientOptions options)
    {
        _options = options;
    }

    public abstract Task<IHttpResponse> ExecuteAsync(IHttpRequest request, CancellationToken cancellationToken = default);

    public virtual async Task<IHttpResponse> GetAsync(string url, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Get;
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual async Task<IHttpResponse> PostAsync(string url, string? body = null, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Post;
        request.Body = body;
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual async Task<IHttpResponse> PostJsonAsync<T>(string url, T content, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Post;
        request.SetJsonBody(content);
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual async Task<IHttpResponse> PutAsync(string url, string? body = null, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Put;
        request.Body = body;
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual async Task<IHttpResponse> PutJsonAsync<T>(string url, T content, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Put;
        request.SetJsonBody(content);
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual async Task<IHttpResponse> DeleteAsync(string url, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Delete;
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual async Task<IHttpResponse> PatchAsync(string url, string? body = null, IReadOnlyDictionary<string, string>? headers = null, CancellationToken cancellationToken = default)
    {
        var request = CreateRequest();
        request.Url = url;
        request.Method = HttpMethod.Patch;
        request.Body = body;
        
        if (headers != null)
        {
            foreach (var header in headers)
            {
                request.SetHeader(header.Key, header.Value);
            }
        }

        return await ExecuteAsync(request, cancellationToken);
    }

    public virtual void SetDefaultTimeout(int timeoutMs)
    {
        _options.DefaultTimeoutMs = timeoutMs;
    }

    public virtual void SetDefaultHeaders(IReadOnlyDictionary<string, string> headers)
    {
        _options.DefaultHeaders.Clear();
        foreach (var header in headers)
        {
            _options.DefaultHeaders[header.Key] = header.Value;
        }
    }

    public virtual void SetClientCertificate(string certPath, string keyPath)
    {
        _options.ClientCertPath = certPath;
        _options.ClientKeyPath = keyPath;
    }

    public virtual void SetCACertificate(string caPath)
    {
        _options.CACertPath = caPath;
    }

    public virtual void SetVerifyPeer(bool verify)
    {
        _options.VerifyPeer = verify;
    }

    public abstract Task<bool> PingAsync(string url, CancellationToken cancellationToken = default);

    public virtual IHttpRequest CreateRequest()
    {
        var request = new HttpRequest
        {
            TimeoutMs = _options.DefaultTimeoutMs,
            VerifyPeer = _options.VerifyPeer,
            FollowRedirects = _options.FollowRedirects,
            ClientCertPath = _options.ClientCertPath,
            ClientKeyPath = _options.ClientKeyPath,
            CACertPath = _options.CACertPath
        };

        // Add default headers
        foreach (var header in _options.DefaultHeaders)
        {
            request.SetHeader(header.Key, header.Value);
        }

        return request;
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (disposing)
            {
                // Dispose managed resources
            }
            _disposed = true;
        }
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }
}

File: Platform/infra/http/interfaces/ts/base-http-client.ts
/**
 * Base HTTP Client Implementation for TypeScript
 * 
 * This module provides the abstract base class for all HTTP client implementations,
 * following the same patterns as the C# version.
 */

import { HttpClient, HttpRequest, HttpResponse, HttpMethod, CoyoteHttpClient } from './http-client.js';
import { HttpClientOptions } from './configuration.js';

/**
 * HTTP request implementation
 */
export class HttpRequestImpl implements HttpRequest {
  url: string = '';
  method: HttpMethod = HttpMethod.GET;
  headers?: Record<string, string> = {};
  body?: string;
  timeout?: number;

  constructor(options?: Partial<HttpRequest>) {
    if (options) {
      Object.assign(this, options);
    }
  }

  setJsonBody<T>(content: T): void {
    this.body = JSON.stringify(content);
    this.setHeader('Content-Type', 'application/json');
  }

  setHeader(name: string, value: string): void {
    if (!this.headers) {
      this.headers = {};
    }
    this.headers[name] = value;
  }
}

/**
 * HTTP response implementation
 */
export class HttpResponseImpl implements HttpResponse {
  statusCode: number = 0;
  body: string = '';
  headers: Record<string, string> = {};
  errorMessage?: string;

  get isSuccess(): boolean {
    return this.statusCode >= 200 && this.statusCode < 300;
  }

  constructor(options?: Partial<HttpResponse>) {
    if (options) {
      Object.assign(this, options);
    }
  }
}

/**
 * Base HTTP client implementation with common functionality
 */
export abstract class BaseHttpClient implements CoyoteHttpClient {
  protected readonly options: HttpClientOptions;
  protected disposed: boolean = false;

  constructor(options: HttpClientOptions) {
    this.options = { ...options };
  }

  abstract executeAsync(request: HttpRequest): Promise<HttpResponse>;
  abstract pingAsync(url: string): Promise<boolean>;

  async getAsync(url: string, headers?: Record<string, string>): Promise<HttpResponse> {
    const request = this.createRequest();
    request.url = url;
    request.method = HttpMethod.GET;
    
    if (headers) {
      Object.assign(request.headers || {}, headers);
    }

    return this.executeAsync(request);
  }

  async postJsonAsync(url: string, data: any, headers?: Record<string, string>): Promise<HttpResponse> {
    const request = this.createRequest();
    request.url = url;
    request.method = HttpMethod.POST;
    request.body = JSON.stringify(data);
    
    const requestHeaders = request.headers || {};
    requestHeaders['Content-Type'] = 'application/json';
    
    if (headers) {
      Object.assign(requestHeaders, headers);
    }
    
    request.headers = requestHeaders;

    return this.executeAsync(request);
  }

  async putJsonAsync(url: string, data: any, headers?: Record<string, string>): Promise<HttpResponse> {
    const request = this.createRequest();
    request.url = url;
    request.method = HttpMethod.PUT;
    request.body = JSON.stringify(data);
    
    const requestHeaders = request.headers || {};
    requestHeaders['Content-Type'] = 'application/json';
    
    if (headers) {
      Object.assign(requestHeaders, headers);
    }
    
    request.headers = requestHeaders;

    return this.executeAsync(request);
  }

  async deleteAsync(url: string, headers?: Record<string, string>): Promise<HttpResponse> {
    const request = this.createRequest();
    request.url = url;
    request.method = HttpMethod.DELETE;
    
    if (headers) {
      Object.assign(request.headers || {}, headers);
    }

    return this.executeAsync(request);
  }

  getContent<T>(response: HttpResponse): T | null {
    if (!response.body) {
      return null;
    }

    try {
      return JSON.parse(response.body) as T;
    } catch (error) {
      console.warn('Failed to parse response body as JSON:', error);
      return null;
    }
  }

  setDefaultTimeout(timeoutMs: number): void {
    this.options.defaultTimeoutMs = timeoutMs;
  }

  setDefaultHeaders(headers: Record<string, string>): void {
    this.options.defaultHeaders = { ...headers };
  }

  setUserAgent(userAgent: string): void {
    this.options.userAgent = userAgent;
  }

  setBaseUrl(baseUrl: string): void {
    this.options.baseUrl = baseUrl;
  }

  createRequest(): HttpRequestImpl {
    const request = new HttpRequestImpl();
    request.timeout = this.options.defaultTimeoutMs;
    request.headers = { ...this.options.defaultHeaders };

    return request;
  }

  protected createUrl(url: string): string {
    if (url.startsWith('http://') || url.startsWith('https://')) {
      return url;
    }

    if (this.options.baseUrl) {
      const baseUrl = this.options.baseUrl.endsWith('/') 
        ? this.options.baseUrl.slice(0, -1) 
        : this.options.baseUrl;
      const path = url.startsWith('/') ? url : `/${url}`;
      return `${baseUrl}${path}`;
    }

    return url;
  }

  protected createResponseHeaders(headers: any): Record<string, string> {
    const result: Record<string, string> = {};
    
    if (headers) {
      // Handle different header formats (Node.js, fetch, etc.)
      if (typeof headers.entries === 'function') {
        // Headers object with entries method
        for (const [name, value] of headers.entries()) {
          result[name.toLowerCase()] = Array.isArray(value) ? value.join(', ') : String(value);
        }
      } else if (typeof headers === 'object') {
        // Plain object
        for (const [name, value] of Object.entries(headers)) {
          if (value !== undefined) {
            result[name.toLowerCase()] = Array.isArray(value) ? value.join(', ') : String(value);
          }
        }
      }
    }

    return result;
  }

  dispose(): void {
    if (!this.disposed) {
      this.disposed = true;
      this.onDispose();
    }
  }

  protected onDispose(): void {
    // Override in derived classes for cleanup
  }
}
